base/shared/src/main/scala/scalaz/data/Option2.scala

```diff
@@ -0,0 +1,12 @@
+package scalaz
+package data
+
+/**
+ * Isomorphic to `Option[(A, B)]`, but avoids allocating a `Tuple2` instance.
+ *
+ * Covariance is intentional. For invariant version, consider implementing `Maybe2`.
+ */
+sealed abstract class Option2[+A, +B]
```

Comment:
Is a covariant `Option2` strictly needed to implement `IsCovariant[IList]`? Seems like you could just implement `IsCovariant2[Maybe2]` or something similar.

Reply:
You could do that, but the problem then shifts to how to implement `IsCovariant2[Maybe2]`. You would have to use `asInstanceOf` or equivalent (e.g. `Liskov.unsafeForce`).

Reply:
Maybe this should trigger a re-discussion of when / how to use variance in Scalaz 8?

Reply:
We could also keep `Option2` package private, and expose `Maybe2` as an invariant type alias for `Option2` (just like `Maybe` in #1450). That way, we safely implement `IsCovariant[Maybe2]` via deferring to covariant `Option2`.

Reply:
I like that, since it means we don't have to make any decisions on variance (defer all the things, the mantra of FP 😆). It would be nice to keep it in the same file to not clutter up the file structure.

Reply:
Aah, we still have to keep `Option2` public in order to support pattern matching on `Maybe2` by conversion to `Option2`.

Reply:
Can't we just have unapply on a `Maybe2` object?

Reply:
We can, but what does it unapply to?

Reply:
A tuple. Which involves additional boxing, sadly.

Reply:
Updated to use `Maybe2`, but still kept `Option2` public.

Reply:
Something like this would work, wouldn't it?

```scala
object Some2 {
  def apply[A, B](v: Maybe2[A, B]): Option[(A, B)] = ???
}
object None2 {
  def apply[A, B](v: Maybe2[A, B]): Option[Unit] = ???
}
```

Then a user can do:

```scala
maybe2 match {
  case Some2(a, b) => ???
  case None2() => ???
}
```

Reply:
You mean `unapply`. Yeah, that would work, but loses exhaustivity checking and allocates a `Tuple2`.

Reply:
Scala is like a hydra: working around one problem you hit another two.

Reply:
I think I found a way to do it:

```
sealed trait Maybe2[A, B]
final case class Some2[A, B](_1: A, _2: B) extends Product2[A, B] with Maybe2[A, B]
final case class None2[A, B]() extends Maybe2[A, B]
object Maybe2 {
  def some[A, B](a: A, b: B): Maybe2[A, B] = new Some2(a, b)
  def none[A, B]: Maybe2[A, B] = new None2[A, B]()
}
object Some2 {
  def unapply[A, B](m: Maybe2[A, B]): Option[Product2[A, B]] = m match {
    case s : Some2[_, _] => Some(s: Product2[A, B])
    case _ => None
  }
}
object None2 {
  def unapply[A, B](m: Maybe2[A, B]): Option[Unit] = m match {
    case s @ Some2(_, _) => None
    case _ => Some(())
  }  
}
Maybe2.some("a", "b") match { case Some2(a, b) => a + b }
```

This does not address exhaustivity. I think there is an SI ticket for that somewhere.

Reply:
Here you would again run into trouble of how to define `IsCovariant`, but you got me thinking in the right direction (minus exhaustivity).